lib/services/embeddings.go

```diff
@@ -0,0 +1,396 @@
+}
+
+// PreciseJSONConfig is the fast JSON marshall/unmarshall configuration
+// we use in utils.FastMarshall() but with more precision on floats. This
+// is required to reduce the lost precision when storing embeddings in the
+// backend.
+var PreciseJSONConfig = jsoniter.Config{
+	EscapeHTML:                    false,
+	MarshalFloatWith6Digits:       false, // keep precision
+	ObjectFieldMustBeSimpleString: true,  // do not unescape object field
+	SortMapKeys:                   true,
+}.Froze()
+
+// MarshalEmbedding marshals the ai.Embedding resource to JSON.
+func MarshalEmbedding(embedding *ai.Embedding) ([]byte, error) {
+	data, err := PreciseJSONConfig.Marshal(embedding)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return data, nil
+}
+
+// UnmarshalEmbedding unmarshals JSON into an ai.Embedding resource.
+func UnmarshalEmbedding(bytes []byte) (*ai.Embedding, error) {
+	if len(bytes) == 0 {
+		return nil, trace.BadParameter("missing embedding data")
+	}
+	var embedding ai.Embedding
+	iter := PreciseJSONConfig.BorrowIterator(bytes)
+	defer PreciseJSONConfig.ReturnIterator(iter)
+
+	iter.ReadVal(&embedding)
+	if iter.Error != nil {
+		return nil, trace.Wrap(iter.Error)
+	}
+
+	return &embedding, nil
+}
+
+// NodeEmbeddingWatcher listen for Node events and asynchronously compute
+// embeddings for known nodes.
+type NodeEmbeddingWatcher struct {
+	*resourceWatcher
+	*nodeEmbeddingCollector
+}
+
+// NewNodeEmbeddingWatcher returns a new instance of NodeEmbeddingWatcher.
+func NewNodeEmbeddingWatcher(ctx context.Context, cfg NodeEmbeddingWatcherConfig) (*NodeEmbeddingWatcher, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	collector := &nodeEmbeddingCollector{
+		NodeEmbeddingWatcherConfig: cfg,
+		initializationC:            make(chan struct{}),
+		currentNodes:               make(map[string]*embeddedNode),
+	}
+	// start the collector as staled.
+	collector.stale.Store(true)
+
+	watcher, err := newResourceWatcher(ctx, collector, cfg.ResourceWatcherConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &NodeEmbeddingWatcher{resourceWatcher: watcher, nodeEmbeddingCollector: collector}, nil
+}
+
+// NodeEmbeddingWatcherConfig is the configuration of the NodeEmbeddingWatcher.
+// It extends the NodeWatcherConfig with a reference to the services.Embeddings.
+// This way, the watcher and collector have access to the embeddings in the
+// backend and can create new embeddings via the ai.Embedder.
+type NodeEmbeddingWatcherConfig struct {
+	NodeWatcherConfig
+	Embeddings
+}
+
+// nodeEmbeddingCollector accompanies resourceWatcher when monitoring currentNodes.
+// It keeps tracks of which node has been embedded and which node requires embedding.
+// The embedding happens asynchronously as calling the openAI API by batch is
+// much quicker, stable and efficient.
+type nodeEmbeddingCollector struct {
```

Comment:
Is this comment a blocker, or can we address it later when we'll extend embeddings? This PR is somewhat important to unlock the rest of the Assist PRs for the next features and it would be easier for us to redesign this later.

I can extract the embedding logic out of the collector, but this will complexify even more an [already complex locking situation](https://github.com/gravitational/teleport/pull/27204#discussion_r1223328498) as the collector and something else would need to sync to ensure embedding consistency. Do you think of a specific design that would allow consistent asynchronous embedding not triggered in the watcher?